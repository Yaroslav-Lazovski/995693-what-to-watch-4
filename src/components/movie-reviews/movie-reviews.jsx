import React, {PureComponent} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";

import MovieReview from "../movie-review/movie-review.jsx";

import {getComments} from "../../reducer/data/selectors";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import {getActiveMovieId} from "../../reducer/state/selectors.js";

export class MovieReviews extends PureComponent {
  componentDidMount() {
    this._loadComments();
  }

  _loadComments() {
    const {getMovieComments, activeMovieId} = this.props;

    getMovieComments(activeMovieId);
  }

  _getReview(review, index) {
    const key = `${review.id}-${index}`;

    return (
      <MovieReview
        key={key}
        review={review}
      />
    );
  }

  renderReviews(columnReviews) {
    return columnReviews.map(this._getReview);
  }

  render() {
    const {reviews} = this.props;

    const halfReviews = Math.ceil(reviews.length / 2);
    const firstColumn = reviews.slice(0, halfReviews);
    const secondColumn = reviews.slice(halfReviews);

    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">
          {this.renderReviews(firstColumn)}
        </div>
        <div className="movie-card__reviews-col">
          {this.renderReviews(secondColumn)}
        </div>
      </div>
    );
  }
}

MovieReviews.propTypes = {
  reviews: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        user: PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired,
        }).isRequired,
        rating: PropTypes.number.isRequired,
        comment: PropTypes.string.isRequired,
        date: PropTypes.string.isRequired
      }).isRequired
  ).isRequired,
  activeMovieId: PropTypes.number.isRequired,
  getMovieComments: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  reviews: getComments(state),
  activeMovieId: getActiveMovieId(state)
});

const mapDispatchToProps = (dispatch) => ({
  getMovieComments(id) {
    dispatch(DataOperation.loadMovieComments(id));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieReviews);
