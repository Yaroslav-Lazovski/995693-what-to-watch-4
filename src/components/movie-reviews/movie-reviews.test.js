import React from "react";
import renderer from "react-test-renderer";

import {MovieReviews} from "./movie-reviews.jsx";


const reviews = [
  {
    id: 1,
    user: {
      id: 4,
      name: `User`
    },
    comment: `Comment`,
    date: `December 24, 2016`,
    rating: 7.1,
  },
  {
    id: 2,
    user: {
      id: 3,
      name: `User`
    },
    comment: `Comment`,
    date: `December 24, 2016`,
    rating: 7.1,
  },
];

const movie = {
  id: 1
};


describe(`MovieReviews Snapshot`, () => {
  it(`Should MovieReviews render correctly`, () => {
    const tree = renderer
      .create(<MovieReviews
        reviews={reviews}
        movie={movie}
        getMovieComments={() => {}}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
