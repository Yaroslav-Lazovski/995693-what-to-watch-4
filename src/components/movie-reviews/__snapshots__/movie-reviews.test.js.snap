// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieReviews Snapshot Should MovieReviews render correctly 1`] = `
<div
  style={
    Object {
      "height": "200px",
      "margin": "0 auto",
      "width": "200px",
    }
  }
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10 50a40 40 0 0080 0 40 42 0 01-80 0"
      fill="#c9b37e"
    >
      <animateTransform
        attributeName="transform"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        type="rotate"
        values="0 50 51;360 50 51"
      />
    </path>
  </svg>
</div>
`;
